File: get-swadesh.py
"""Scrape Swadesh word lists from Wiktionary and save to JSON.

Crawls the Wiktionary Swadesh list category, parses each language's
table into English-keyed word entries, and writes the full dataset
to ``swadesh_list_2.json``.
"""

import requests
from bs4 import BeautifulSoup
import json
import re
import unicodedata
from concurrent.futures import ThreadPoolExecutor, as_completed

BASE = "https://en.wiktionary.org"
CATEGORY_URL = BASE + "/wiki/Category:Swadesh_lists_by_language"
HEADERS = {"User-Agent": "SwadeshScraper/1.0"}


def clean_text(element, is_language_header=False):
    """Normalize raw HTML cell content into a clean Unicode string.

    Strips newlines, citation brackets, and edit links. When
    ``is_language_header`` is True, non-Latin characters are dropped so
    that column headers contain only ASCII/extended-Latin text.

    Args:
        element: A BeautifulSoup tag, a plain string, or ``None``.
        is_language_header: If True, strip non-Latin characters.

    Returns:
        A NFC-normalized, whitespace-collapsed string, or ``""`` if the
        input is falsy.
    """
    if not element:
        return ""
    text = element if isinstance(element, str) else element.get_text(separator=" ")
    text = text.replace('\n', ' ').replace('\r', ' ').replace('\t', ' ')
    # Remove "edit (N)" links injected by MediaWiki
    text = re.sub(r'edit\s*\(\d+\)', '', text)
    # Remove footnote references like [1], [2]
    text = re.sub(r'\[\d+\]', '', text)
    if is_language_header:
        # Keep only Latin-script characters and basic punctuation
        latin_only = re.findall(r'[A-Za-z\u00C0-\u024F\s\-\(\)]+', text)
        text = "".join(latin_only)
        text = text.replace('()', '').strip()
    # Tidy up stray whitespace around punctuation
    text = re.sub(r'\(\s+', '(', text)
    text = re.sub(r'\s+\)', ')', text)
    text = re.sub(r'\s+([*,;:?\.])', r'\1', text)
    text = unicodedata.normalize('NFC', text)
    text = re.sub(r'\s+', ' ', text).strip()
    return text


def get_swadesh_pages():
    """Return all Swadesh list page URLs and their language names.

    Paginates through the Wiktionary category until no "next page" link
    is found.

    Returns:
        A deduplicated list of ``(url, language_name)`` tuples.
    """
    pages = []
    url = CATEGORY_URL
    while url:
        print(f"Crawling Category Page: {url}")
        r = requests.get(url, headers=HEADERS)
        soup = BeautifulSoup(r.text, "html.parser")
        for a in soup.select(".mw-category a"):
            href = a.get("href", "")
            if href.startswith("/wiki/Appendix:") and "Swadesh" in href:
                lang_name = a.get_text().replace("Appendix:", "").split("Swadesh")[0].strip()
                pages.append((BASE + href, lang_name))
        next_link = soup.find("a", string=re.compile(r"next (page|200)", re.I))
        url = BASE + next_link["href"] if next_link and "href" in next_link.attrs else None
    return list(set(pages))


def parse_page(url, lang_name):
    """Parse a single Wiktionary Swadesh list page into structured entries.

    Locates the first ``wikitable`` on the page, identifies the English
    column and any non-blacklisted target-language columns, then yields
    one dict per row mapping ``"english"`` and column-name keys to cell
    text.

    Columns whose headers match terms like "ipa", "notes", "cognate", etc.
    are skipped to avoid including metadata in word entries.

    Args:
        url: Full URL of the Wiktionary appendix page.
        lang_name: Human-readable language name used as the ``"language"``
            field in the returned data.

    Returns:
        A list containing a single ``{"language": ..., "entries": [...]}``
        dict, or an empty list if no usable table was found.
    """
    r = requests.get(url, headers=HEADERS)
    r.encoding = 'utf-8'
    soup = BeautifulSoup(r.text, "html.parser")

    table = soup.find("table", class_="wikitable")
    if not table:
        return []

    rows = table.find_all("tr")
    if not rows:
        return []

    header_cells = rows[0].find_all(["th", "td"])
    col_names = [clean_text(c) for c in header_cells]

    # Column header fragments that indicate metadata rather than word forms
    BLACKLIST = [
        "no.", "№", "number", "example", "term?",
        "ipa", "transcription", "phonetic", "transliteration",
        "notes", "gloss", "reconstruction", "cognate", "literal"
    ]

    idx_eng = -1
    target_indices = []

    for i, name in enumerate(col_names):
        name_low = name.lower()
        if "english" in name_low:
            idx_eng = i
            continue
        if not any(term in name_low for term in BLACKLIST):
            target_indices.append(i)

    # Fall back to column 1 if no explicit English header was found
    if idx_eng == -1:
        idx_eng = 1

    entries = []
    for row in rows[1:]:
        cols = row.find_all(["td", "th"])
        max_idx = max(idx_eng, max(target_indices, default=0))
        if len(cols) <= max_idx:
            continue

        english_val = clean_text(cols[idx_eng])
        if not english_val:
            continue

        entry = {"english": english_val}
        for idx in target_indices:
            if idx >= len(cols):
                continue
            word_val = clean_text(cols[idx])
            # Skip cells that are clearly empty or placeholder dashes
            if not word_val or word_val in ["-", "-", "?", "？"]:
                continue
            key = col_names[idx].lower() if col_names[idx] else f"col_{idx}"
            entry[key] = word_val

        # Only keep entries that have at least one non-English word
        if len(entry) > 1:
            entries.append(entry)

    return [{"language": lang_name, "entries": entries}] if entries else []


def main():
    """Scrape all Swadesh pages and write results to ``swadesh_list_2.json``."""
    pages = get_swadesh_pages()
    print(f"Found {len(pages)} pages to scrape.")

    all_data = []

    def scrape(i, url, lang_name):
        """Fetch and parse one page; print a progress line on success."""
        entries = parse_page(url, lang_name)
        if entries:
            print(f"   [{i+1}/{len(pages)}] {lang_name} - {entries[0]['entries'][0]}")
        return entries

    with ThreadPoolExecutor(max_workers=8) as executor:
        futures = {
            executor.submit(scrape, i, url, lang_name): url
            for i, (url, lang_name) in enumerate(pages)
        }
        for future in as_completed(futures):
            try:
                entries = future.result()
                if entries:
                    all_data.extend(entries)
            except Exception as e:
                print(f"   Error: {e}")

    output_filename = "swadesh_list.json"
    with open(output_filename, "w", encoding="utf-8") as f:
        json.dump(all_data, f, ensure_ascii=False, indent=2)

    print(f"\nDone! Saved {len(all_data)} entries to {output_filename}")


if __name__ == "__main__":
    main()

File: get-swadesh.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"regexp"
	"strings"
	"sync"
	"unicode"

	"golang.org/x/net/html"
	"golang.org/x/text/unicode/norm"
)

const BASE = "https://en.wiktionary.org"

var scraperHeader = http.Header{
	"User-Agent": []string{"SwadeshScraper/1.0"},
}

var httpClient = &http.Client{}

func CreateSwadeshJson() {
	pages := GetSwadeshPages()
	fmt.Printf("Found %v pages to scrape\n", len(pages))

	// create vars to store data and for concurrency stuff
	var (
		allData []map[string]any
		mu      sync.Mutex
		wg      sync.WaitGroup
		sem     = make(chan struct{}, 8) // max 8 concurrent workers
	)

	for i, page := range pages {
		wg.Add(1)
		go func(i int, url, langName string) {
			defer wg.Done()

			sem <- struct{}{}        // sends an empty struct to the channel
			defer func() { <-sem }() // goroutine finishes, calls anonymous func, removes goroutine

			ld := ParsePage(url, langName)
			if len(ld.Entries) > 0 {
				// for checking progress
				fmt.Printf("   [%v/%v] %s - %v\n", i+1, len(pages), langName, ld.Entries[0])

				// lock for writing data
				mu.Lock()
				allData = append(allData, map[string]any{
					"language": ld.LangName,
					"entries":  ld.Entries,
				})
				mu.Unlock()
			}
		}(i, page.URL, page.LangName)
	}
	wg.Wait()

	outputFilename := "swadesh_list.json"
	f, err := os.Create(outputFilename)
	if err != nil {
		fmt.Printf("Error creating output file: %v\n", err)
		return
	}
	defer f.Close()

	enc := json.NewEncoder(f)
	enc.SetIndent("", "  ")
	enc.SetEscapeHTML(false)
	if err := enc.Encode(allData); err != nil {
		fmt.Printf("Error encoding JSON: %v\n", err)
		return
	}

	fmt.Printf("\nDone! Saved %v entries to %s\n", len(allData), outputFilename)
}

// CleanText normalizes raw HTML cell content into a clean Unicode string.
func CleanText(element string, isLanguageHeader bool) string {
	if element == "" {
		return ""
	}

	text := element
	text = strings.NewReplacer("\n", " ", "\r", " ", "\t", " ").Replace(text)

	// Remove "edit(123)" artifacts from wiktionary
	editRe := regexp.MustCompile(`edit\s*\(\d+\)`)
	text = editRe.ReplaceAllString(text, "")

	if isLanguageHeader {
		// Keep only Latin-range characters (basic + extended), spaces, hyphens, parens
		latinRe := regexp.MustCompile(`[A-Za-z\x{00C0}-\x{024F}\s\-()\[\]]+`)
		parts := latinRe.FindAllString(text, -1)
		text = strings.Join(parts, "")
		text = strings.ReplaceAll(text, "()", "")
		text = strings.TrimSpace(text)
	}

	// Fix spacing around punctuation
	text = regexp.MustCompile(`\(\s+`).ReplaceAllString(text, "(")
	text = regexp.MustCompile(`\s+\)`).ReplaceAllString(text, ")")
	text = regexp.MustCompile(`\s+([*,;:?.])`).ReplaceAllString(text, "$1")

	// NFC normalization
	text = norm.NFC.String(text)

	// Collapse multiple spaces
	text = regexp.MustCompile(`\s+`).ReplaceAllString(text, " ")
	text = strings.TrimSpace(text)

	return text
}

type Page struct {
	URL      string
	LangName string
}

// GetSwadeshPages crawls the Wiktionary Swadesh list category and returns all pages.
func GetSwadeshPages() []Page {
	var pages []Page
	seen := map[string]bool{}

	url := BASE + "/wiki/Category:Swadesh_lists_by_language"

	nextPageRe := regexp.MustCompile(`(?i)next\s+(page|200)`)

	for url != "" {
		fmt.Printf("Crawling Category Page: %s\n", url)

		doc, err := fetchHTML(url)
		if err != nil {
			fmt.Printf("Error fetching %s: %v\n", url, err)
			break
		}

		// Find all <a> tags inside .mw-category
		var nextURL string
		var walk func(*html.Node, bool)
		walk = func(n *html.Node, inCategory bool) {
			if n.Type == html.ElementNode && n.Data == "div" {
				for _, a := range n.Attr {
					if a.Key == "class" && strings.Contains(a.Val, "mw-category") {
						inCategory = true
					}
				}
			}
			if n.Type == html.ElementNode && n.Data == "a" {
				href := attrVal(n, "href")
				text := extractText(n)
				if inCategory && strings.HasPrefix(href, "/wiki/Appendix:") && strings.Contains(href, "Swadesh") {
					langName := strings.TrimSpace(strings.Split(strings.ReplaceAll(text, "Appendix:", ""), "Swadesh")[0])
					fullURL := BASE + href
					if !seen[fullURL] {
						seen[fullURL] = true
						pages = append(pages, Page{URL: fullURL, LangName: langName})
					}
				}
				// Detect "next page" / "next 200" link
				if nextPageRe.MatchString(text) {
					if href != "" {
						nextURL = BASE + href
					}
				}
			}
			for c := n.FirstChild; c != nil; c = c.NextSibling {
				walk(c, inCategory)
			}
		}
		walk(doc, false)

		url = nextURL
	}

	return pages
}

type LanguageData struct {
	LangName string
	Entries  []map[string]string
}

var blacklist = []string{
	"no", "no.", "№", "number", "example", "term?",
	"ipa", "transcription", "phonetic", "transliteration",
	"notes", "gloss", "reconstruction", "cognate", "literal",
}

// ParsePage parses a single Wiktionary Swadesh list page into structured entries.
func ParsePage(url, langName string) LanguageData {
	doc, err := fetchHTML(url)
	if err != nil {
		fmt.Printf("Error fetching %s: %v\n", url, err)
		return LanguageData{}
	}

	// Find the first wikitable
	table := findNode(doc, func(n *html.Node) bool {
		return n.Type == html.ElementNode && n.Data == "table" &&
			strings.Contains(attrVal(n, "class"), "wikitable")
	})
	if table == nil {
		return LanguageData{}
	}

	rows := findNodes(table, func(n *html.Node) bool {
		return n.Type == html.ElementNode && n.Data == "tr"
	})
	if len(rows) == 0 {
		return LanguageData{}
	}

	// Parse header row
	headerCells := findNodes(rows[0], func(n *html.Node) bool {
		return n.Type == html.ElementNode && (n.Data == "th" || n.Data == "td")
	})
	colNames := make([]string, len(headerCells))
	for i, cell := range headerCells {
		colNames[i] = CleanText(extractText(cell), true)
	}

	idxEng := -1
	var targetIndices []int
	for i, name := range colNames {
		nameLow := strings.ToLower(name)
		if strings.Contains(nameLow, "english") {
			idxEng = i
			continue
		}
		isBlacklisted := name == ""
		for _, term := range blacklist {
			if strings.Contains(nameLow, term) {
				isBlacklisted = true
				break
			}
		}
		if !isBlacklisted {
			targetIndices = append(targetIndices, i)
		}
	}
	if idxEng == -1 {
		idxEng = 1
	}

	var entries []map[string]string
	for _, row := range rows[1:] {
		cols := findNodes(row, func(n *html.Node) bool {
			return n.Type == html.ElementNode && (n.Data == "td" || n.Data == "th")
		})

		maxIdx := idxEng
		for _, idx := range targetIndices {
			if idx > maxIdx {
				maxIdx = idx
			}
		}
		if len(cols) <= maxIdx {
			continue
		}

		englishVal := CleanText(extractText(cols[idxEng]), false)
		if englishVal == "" {
			continue
		}

		entry := map[string]string{"english": englishVal}
		for _, idx := range targetIndices {
			if idx >= len(cols) {
				continue
			}
			wordVal := CleanText(extractText(cols[idx]), false)
			if wordVal == "" || wordVal == "-" || wordVal == "–" || wordVal == "?" || wordVal == "？" {
				continue
			}
			key := fmt.Sprintf("col_%d", idx)
			if colNames[idx] != "" {
				key = strings.ToLower(colNames[idx])
			}
			entry[key] = wordVal
		}

		if len(entry) > 1 {
			entries = append(entries, entry)
		}
	}

	return LanguageData{LangName: langName, Entries: entries}
}

// --- HTML helpers ---

func fetchHTML(url string) (*html.Node, error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}
	for key, vals := range scraperHeader {
		for _, v := range vals {
			req.Header.Set(key, v)
		}
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	return html.Parse(strings.NewReader(string(body)))
}

func attrVal(n *html.Node, key string) string {
	for _, a := range n.Attr {
		if a.Key == key {
			return a.Val
		}
	}
	return ""
}

func extractText(n *html.Node) string {
	var sb strings.Builder
	var walk func(*html.Node)
	walk = func(n *html.Node) {
		if n.Type == html.TextNode {
			sb.WriteString(n.Data)
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(n)
	// Remove non-printable characters
	return strings.Map(func(r rune) rune {
		if unicode.IsPrint(r) || r == ' ' {
			return r
		}
		return -1
	}, sb.String())
}

func findNode(n *html.Node, match func(*html.Node) bool) *html.Node {
	if match(n) {
		return n
	}
	for c := n.FirstChild; c != nil; c = c.NextSibling {
		if found := findNode(c, match); found != nil {
			return found
		}
	}
	return nil
}

func findNodes(n *html.Node, match func(*html.Node) bool) []*html.Node {
	var results []*html.Node
	var walk func(*html.Node)
	walk = func(n *html.Node) {
		if match(n) {
			results = append(results, n)
			return // don't descend into matched node (e.g. nested rows)
		}
		for c := n.FirstChild; c != nil; c = c.NextSibling {
			walk(c)
		}
	}
	walk(n)
	return results
}
